core/drivers/hisi_trng.c

```diff
@@ -0,0 +1,124 @@
+// SPDX-License-Identifier: GPL-2.0
+/* Copyright (c) 2021 HiSilicon Limited. */
+
+#include <initcall.h>
+#include <io.h>
+#include <malloc.h>
+#include <mm/core_memprot.h>
+#include <platform_config.h>
+#include <rng_support.h>
+
+#define TRNG_MODULE_VERSION "22.12.1"
+#define HTRNG_RANDATA_REG 0xF0
+#define HTRNG_ALG_MODE_REG 0x18C
+#define HTRNG_DATA_ST_REG 0x1B4
+#define HTRNG_VERSION_REG 0x1B8
+#define HTRNG_VERSION_V2 GENMASK_32(31, 0)
+#define HTRNG_VERSION_V3 0x30
+#define HTRNG_DRBG_ALG_SM4 BIT(1)
+#define HTRNG_RANDATA_ERR_ST GENMASK_32(15, 1)
+#define HTRNG_BYTES 4
+
+struct hisi_trng {
+	uintptr_t base;
+	uint32_t ver;
+};
+
+union trng_fifo {
+	uint32_t val;
+	uint8_t data_byte[HTRNG_BYTES];
+};
+
+static struct hisi_trng *trng_dev;
+
+static TEE_Result trng_read(uint32_t *val)
+{
+	uint32_t status;
+
+	*val = io_read32(trng_dev->base + HTRNG_RANDATA_REG);
+	if (*val == 0) {
+		if (trng_dev->ver == HTRNG_VERSION_V3) {
+			status = io_read32(trng_dev->base + HTRNG_DATA_ST_REG);
+			if (status & HTRNG_RANDATA_ERR_ST) {
+				EMSG("TRNG hardware error 0x%x happened!\n",
+				     status);
+				return TEE_ERROR_BAD_STATE;
+			}
+		} else {
+			EMSG("TRNG hardware 0 data error!\n");
+			return TEE_ERROR_BAD_STATE;
+		}
+	}
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result hw_get_random_bytes(void *buf, size_t len)
+{
+	size_t buf_pos = 0, fifo_pos = 0;
+	uint8_t *out = (uint8_t *)buf;
+	union trng_fifo fifo = { };
```

Comment:
Could this be static to save a partial read between calls? Compare with other implementations of this function.